main.py

```diff
@@ -119,61 +127,97 @@ def update():
-            flowobj.__dict__[key] = st.number_input(f"{key}", value=float(val), key=f"{id}_{key}_{flowobj}")
-    return
-
-## TODO: ASK SIMON TO ANNOTATE BELOW
 ## Add element sidebar tab
 with add_element:
+    # Get user input on what element to add
     key = st.selectbox("Select Flow Element", options=ELEMENT_DEFAULT_DICT.keys())
 
-    try:
-        flowobj = copy.deepcopy(ELEMENT_DEFAULT_DICT[key])
-        name    = adjust_objects([flowobj], "element")
-        if st.button("Add ", key="add_element"):
-            st.session_state["field"].objects.extend([flowobj])
+    # Retrieve necessary arguments for class using default values from a default element
+    proto_elem  = ELEMENT_DEFAULT_DICT[key]
+    args        = [None]*len(proto_elem.__dict__)
+    # Create input fields
+    for i, (k, v) in enumerate(proto_elem.__dict__.items()):
+        # usually strength has some condition, i.e. Sources / Sinks are defined by their sign, so we add case studies
+        if k == 'strength':
+            if   flow_element_type(proto_elem) == 'Source':
+                args[i] = st.number_input(f"{k}", value=float(v), min_value= 0.01, key=f"addelement_{key}_{i}")
+            elif flow_element_type(proto_elem) == 'Sink':
+                args[i] = st.number_input(f"{k}", value=float(v), max_value=-0.01, key=f"addelement_{key}_{i}")
+        # otherwise it is inputs
+        else:
+            args[i] = st.number_input(f"{k}", value=float(v), key=f"addelement_{key}_{i}")
+
+    # Add element
+    if st.button("Add", key="add_element"):
+        add_authority = True
+
+        ## Ensure Source / Sink overlap does not happen, as it causes an issue with ff.create_streamline()
+        if flow_element_type(proto_elem) == 'Source' or flow_element_type(proto_elem) == 'Sink':
+            keys = list(st.session_state["field"].objects.keys())
+            for key in keys:
+                elem = st.session_state["field"].objects[key]
+                # Compare arguments with all other sources/sink
+                if flow_element_type(elem) == 'Source' or flow_element_type(elem) == 'Sink':
+                    x0 = args[0]; y0 = args[1]                # Wish it could be more elegant, but this works.
+                    x1 = elem.x;  y1 = elem.y
+
+                    if abs(x0-x1) < 0.01 or abs(y0-y1) < 0.01:
+                        add_authority = False
+
+        ## Add item to flowfield dictionary
+        if add_authority:
+            elem = proto_elem.__class__
+            num  = len(st.session_state["field"].objects) + 1
+            name = f"{num}. [{flow_element_type(proto_elem)}]"
+            st.session_state["field"].objects[name] = elem(*args)
+
+            st.markdown(f'Added {name}')
             update()
 
-    except KeyError:
-        pass
+        else:
+            st.markdown(f'Did not add element -- check that Source/Sink position does not conflict with already existing Source/Sink positions')
 
```

Comment:
Nice fix for now, but this is not a permanent solution, right? It shouldn't crash in these situations, as a doublet could be created in this way.

I think a way to solve this could be to merge existing objects. This might still cause user issues though.

Reply:
That might be something interesting to look at...

Reply:
Will add as a feature that can be implemented into the list